cmd/docker-proxy/main.go

```diff
@@ -13,11 +13,27 @@ import (
 	"github.com/ishidawataru/sctp"
 )
 
+// The caller is expected to pass-in open file descriptors ...
+const (
+	// Pipe for reporting status, as a string. "0\n" if the proxy
+	// started normally. "1\n<error message>" otherwise.
+	parentPipeFd uintptr = 3 + iota
+	// If -use-listen-fd=true, a listening socket ready to accept TCP
+	// connections or receive UDP. (Without that option on the command
+	// line, the listener needs to be opened by docker-proxy, for
+	// compatibility with older docker daemons. In this case fd 4
+	// may belong to the Go runtime.)
+	listenSockFd
+)
+
 func main() {
-	f := os.NewFile(3, "signal-parent")
-	host, container := parseFlags()
+	config := parseFlags()
+	p, err := newProxy(config)
+	if config.ListenSock != nil {
+		config.ListenSock.Close()
+	}
 
-	p, err := NewProxy(host, container)
+	f := os.NewFile(parentPipeFd, "signal-parent")
```

Comment:
While we're on the topic of non-blocking fds, I am tempted to set the parentPipeFd to non-blocking mode. It is a premature optimization, I admit -- we will never fill up the kernel's pipe buffer with how little we are writing today -- but it's so trivial to implement, and so obviously correct and risk-free! And if other people are going to use our code as an example to copy from, they'd end up with an implementation that is free from surprising performance characteristics.
```suggestion
	_ = runtime.SetNonblock(parentPipeFd, true)
	f := os.NewFile(parentPipeFd, "signal-parent")
```

Reply:
Ok